examples/joke_bot/src/main.rs

```diff
@@ -0,0 +1,268 @@
+                Ok(_) => {}
+                Err(err) => println!("failed to accept invitation for room {}: {}", &room_id, err),
+            }
+        }
+    }
+    Ok(())
+}
+
+async fn handle_messages(
+    http_client: &HttpClient,
+    matrix_client: &MatrixClient,
+    e: &Raw<AnySyncRoomEvent>,
+    room_id: &RoomId,
+    user_id: &UserId,
+) -> Result<(), Box<dyn Error>> {
+    if let Ok(AnySyncRoomEvent::Message(AnySyncMessageEvent::RoomMessage(m))) = e.deserialize() {
+        // workaround because Conduit does not implement filtering.
+        if &m.sender == user_id {
+            return Ok(());
+        }
+
+        if let MessageType::Text(t) = m.content.msgtype {
+            println!("{}:\t{}", m.sender, t.body);
+            if t.body.to_ascii_lowercase().contains("joke") {
+                let joke = match get_joke(http_client).await {
+                    Ok(joke) => joke,
+                    Err(_) => "I thought of a joke... but I just forgot it.".to_owned(),
+                };
+                let joke_content = RoomMessageEventContent::text_plain(joke);
+
+                let txn_id = generate_txn_id();
+                let req = send_message_event::Request::new(room_id, &txn_id, &joke_content)?;
+                // Just bail if we can't send the message.
+                let _ = matrix_client.send_request(req).await;
+            }
+        }
+    }
+    Ok(())
+}
+
+async fn handle_invitations(
+    http_client: &HttpClient,
+    matrix_client: &MatrixClient,
+    room_id: &RoomId,
+) -> Result<(), Box<dyn Error>> {
+    println!("invited to {}", &room_id);
+    matrix_client
+        .send_request(ruma::api::client::r0::membership::join_room_by_id::Request::new(room_id))
+        .await?;
+
+    let greeting = "Hello! My name is Mr. Bot! I like to tell jokes. Like this one: ";
+    let joke = get_joke(http_client).await.map_or_else(|_| "err... never mind.".to_owned(), |j| j);
+    let content = RoomMessageEventContent::text_plain(format!("{}\n{}", greeting, joke));
+    let txn_id = generate_txn_id();
+    let message = send_message_event::Request::new(room_id, &txn_id, &content)?;
+    matrix_client.send_request(message).await?;
+    Ok(())
+}
+
+// Each message needs a unique transaction ID, otherwise the server thinks that the message is
+// being retransmitted. We could use a random value, or a counter, but to avoid having to store
+// the state, we'll just use the current time as a transaction ID.
+fn generate_txn_id() -> String {
+    SystemTime::now()
+        .duration_since(SystemTime::UNIX_EPOCH)
+        .expect("current time is earlier than Unix epoch")
+        .as_millis()
+        .to_string()
+}
+
+async fn get_joke(client: &HttpClient) -> Result<String, Box<dyn Error>> {
+    let uri = "https://v2.jokeapi.dev/joke/Programming,Pun,Misc?safe-mode&type=single"
+        .parse::<hyper::Uri>()?;
+    let rsp = client.get(uri).await?;
```

Comment:
I think the `get` function would accept a `&str` as well, no? Do you really have to `.parse()` it first?

Reply:
Just using a string for me doesn't compile: https://docs.rs/hyper/latest/hyper/client/struct.Client.html#method.get. :man_shrugging:

Reply:
Oh, I guess I was thinking of `reqwest`s API.